source/solvers/mf_navier_stokes.cc

```diff
@@ -44,6 +45,209 @@
 
 #include <deal.II/numerics/vector_tools.h>
 
+template <typename VectorType>
+class PreconditionBase
+{
+public:
+  virtual void
+  vmult(VectorType &dst, const VectorType &src) const = 0;
+};
+
+template <typename VectorType>
+class MyDiagonalMatrix : public PreconditionBase<VectorType>
```

Comment:
Same as above, won't repeat it for the other classes.
MyDiagonalMatrix seems a weird class name choice. Maybe PreconditionDiagonalMatrix?